### PR TITLE
Upgrade to React 18

### PR DESCRIPTION
**Pull request by tbezman**
Actually turned out not being so crazy.

- Looks like we had a bunch of fucked types in `useCallback` that are now fixed.
- Had to force a resolution on react types since third party libs were bringing in react 17 types.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -13,6 +13,9 @@
     "cypress:gh": "synpress run --record --key 9d633f22-7630-4a11-9cf4-ff9809332564 --configFile cypress.config.ts --group Desktop -b chromium",
     "find-unused-modules": "ts-prune -s .graphql.ts -i pages/"
   },
+  "resolutions": {
+    "@types/react": "^18.0.0"
+  },
```
Comment (tbezman) [inline package.json:18]: Need to make sure any third party libs aren't bringing us down to the React 17 types.

File: pages/_app.tsx [hunk 2]
```diff
@@ -25,9 +25,15 @@ export type MetaTagProps = {
...
-);
+function SafeHydrate({ children }: PropsWithChildren) {
+  const [render, setRender] = useState(false);
+
+  useEffect(() => {
+    setRender(true);
+  }, []);
+
+  return <div suppressHydrationWarning>{render ? children : null}</div>;
+}
```
Comment (tbezman) [inline pages/_app.tsx:36]: Achieves the same thing but this is the React approved way of doing it

File: package.json [hunk 3]
```diff
@@ -49,8 +51,8 @@
     "next": "^12.3.0",
     "nextjs-routes": "^0.1.5",
     "qr-code-styling": "^1.6.0-rc.1",
-    "react": "^17.0.2",
-    "react-dom": "^17.0.2",
+    "react": "18",
+    "react-dom": "18",
```
Comment (Robinnnnn) [inline package.json:55]: should we lock down a more specific version?

File: pages/_app.tsx (hunk 2 above)
Comment (Robinnnnn) [reply pages/_app.tsx:36]: can we add a comment explaining this

File: src/contexts/boundary/ErrorBoundary.tsx [hunk 4]
```diff
@@ -1,12 +1,12 @@
-import { Component } from 'react';
+import { Component, PropsWithChildren } from 'react';
```
Comment (Robinnnnn) [inline src/contexts/boundary/ErrorBoundary.tsx:1]: `PropsWithChildren` is clutchhh

**Review by Robinnnnn — APPROVED** [review]
let's gooooo!

File: package.json (hunk 1 above)
Comment (Robinnnnn) [reply package.json:18]: fantastic catch

Comment (jakzaizzat) [reply package.json:18]: 🧠

**Review by jakzaizzat — APPROVED** [review]
LFG!

**Review by kaitoo1 — APPROVED** [review]

**Pull request merged**